modules/dreamview/frontend/src/renderer/tileground.js

```diff
@@ -58,15 +80,44 @@ export default class TileGround {
                 object.material.dispose();
             }
         }
-        delete this.currentTiles[key];
+        delete items[key];
     }
 
-    appendTiles(row, col, key, coordinates, scene) {
-        if (col < 0 || col > this.metadata.numCols ||
-            row < 0 || row > this.metadata.numRows) {
-            return;
-        }
+    appendItems(row, col, key, coordinates, scene, itemInfo, currentItems,
+        material, object, scales) {
+        itemInfo.forEach((item) => {
+            if (isNaN(item.x) || isNaN(item.y) || isNaN(item.heading)) {
+                return;
+            }
+            const { i, j } = this.getRowAndCol(item.x, item.y);
+            if (i !== row || j !== col) {
+                return;
+            }
+            const itemPos = coordinates.applyOffset({
+                x: item.x,
+                y: item.y,
+                z: 0,
+            });
+            loadObject(material, object, scales,
+                mesh => {
+                    mesh.rotation.x = Math.PI / 2;
+                    mesh.rotation.y = item.heading;
+                    mesh.position.set(itemPos.x, itemPos.y, itemPos.z);
+                    if (item.id) {
+                        // store signal id as name in mesh
+                        mesh.name = item.id;
+                    }
+                    mesh.matrixAutoUpdate = false;
+                    mesh.updateMatrix();
+
+                    // a tile may have multiple signal or stopsign
+                    currentItems[`${key}_${item.x}_${item.y}`] = mesh;
```

Comment:
How come we want to include x,y? Key is not unique?

Reply:
For same row and col, multiple signals can exists, thus add x,y for key.